src/libraries/Native/Unix/System.Security.Cryptography.Native/pal_evp_cipher.c

```diff
@@ -4,10 +4,56 @@
 #include "pal_evp_cipher.h"
 
 #include <assert.h>
+#include <string.h>
 
 #define SUCCESS 1
 #define KEEP_CURRENT_DIRECTION -1
 
+// With some versions of OpenSSL on some IBM Z models, when
+// hardware-accelerated cryptography routines are enabled,
+// calls to EVP_CipherInit_ex may have no effect unless a key
+// is provided.  To work around this problem, we pass any key
+// that was provided earlier to all subsequent EVP_CipherInit_ex
+// calls as well.  This requires saving that key in an extra
+// buffer, which we stash into the EVP_CIPHER_CTX using the
+// "app_data" mechanism.
+#ifdef __s390x__
```

Comment:
nit: I'd use `if defined(TARGET_S390X)` that we're using elsewhere rather than the compiler define.